src/fu-self-test.c

```diff
@@ -1998,6 +1998,89 @@ fu_engine_history_inherit (gconstpointer user_data)
+	g_autoptr(FuDevice) device = fu_device_new_with_context (ctx);
+	g_autoptr(FuEngine) engine = fu_engine_new (FU_APP_FLAGS_NONE);
+	g_autoptr(FuInstallTask) task = NULL;
+	g_autoptr(GBytes) blob_cab = NULL;
+	g_autoptr(GError) error = NULL;
+	g_autoptr(GPtrArray) devices = NULL;
+	g_autoptr(XbNode) component = NULL;
+	g_autoptr(XbSilo) silo_empty = xb_silo_new ();
+	g_autoptr(XbSilo) silo = NULL;
+
+	/* no metadata in daemon */
+	fu_engine_set_silo (engine, silo_empty);
+
+	/* set up dummy plugin */
+	g_setenv ("FWUPD_PLUGIN_TEST", "fail", TRUE);
+	fu_engine_add_plugin (engine, self->plugin);
+	g_setenv ("CONFIGURATION_DIRECTORY", TESTDATADIR_SRC, TRUE);
+	ret = fu_engine_load (engine, FU_ENGINE_LOAD_FLAG_NONE, &error);
+	g_assert_no_error (error);
+	g_assert (ret);
+	g_assert_cmpint (fu_engine_get_status (engine), ==, FWUPD_STATUS_IDLE);
+
+	/* add a device so we can get upgrade it */
+	fu_device_set_version_format (device, FWUPD_VERSION_FORMAT_TRIPLET);
+	fu_device_set_version (device, "1.2.2");
+	fu_device_set_id (device, "test_device");
+	fu_device_add_vendor_id (device, "USB:FFFF");
+	fu_device_add_protocol (device, "com.acme");
+	fu_device_set_name (device, "Test Device");
+	fu_device_set_plugin (device, "test");
+	fu_device_add_guid (device, "12345678-1234-1234-1234-123456789012");
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_set_created (device, 1515338000);
+	fu_engine_add_device (engine, device);
+	devices = fu_engine_get_devices (engine, &error);
+	g_assert_no_error (error);
+	g_assert (devices != NULL);
+	g_assert_cmpint (devices->len, ==, 1);
+	g_assert (fu_device_has_flag (device, FWUPD_DEVICE_FLAG_REGISTERED));
+
+	filename = g_build_filename (TESTDATADIR_DST, "missing-hwid", "noreqs-1.2.3.cab", NULL);
+	blob_cab = fu_common_get_contents_bytes	(filename, &error);
+	g_assert_no_error (error);
+	g_assert (blob_cab != NULL);
+	silo = fu_engine_get_silo_from_blob (engine, blob_cab, &error);
+	g_assert_no_error (error);
+	g_assert_nonnull (silo);
+
+	/* get component */
+	component = xb_silo_query_first (silo, "components/component/id[text()='com.hughski.test.firmware']/..", &error);
+	g_assert_no_error (error);
+	g_assert_nonnull (component);
+
+	/* install it */
+	g_setenv ("FWUPD_PLUGIN_TEST", "requires-reboot", TRUE);
+	task = fu_install_task_new (device, component);
+	ret = fu_engine_install (engine, task, blob_cab,
+				 FWUPD_INSTALL_FLAG_NONE,
+				 FWUPD_FEATURE_FLAG_NONE,
+				 &error);
+	g_assert_no_error (error);
+	g_assert (ret);
+
+	/* check the device requires an activation */
+	g_assert_true (fu_device_has_flag (device, FWUPD_DEVICE_FLAG_NEEDS_REBOOT));
+	g_assert_cmpint (fu_device_get_update_state (device), ==, FWUPD_UPDATE_STATE_NEEDS_REBOOT);
+	g_assert_cmpstr (fu_device_get_version (device), ==, "1.2.2");
+
+	/* delete files */
+	localstatedir = fu_common_get_path (FU_PATH_KIND_LOCALSTATEDIR_PKG);
+	history_db = g_build_filename (localstatedir, "pending.db", NULL);
+	g_unlink (history_db);
```

Comment:
Can you avoid even having a history database by using `FU_ENGINE_LOAD_FLAG_READONLY`?

Reply:
If I add the flag to the engine and remove the deletion, somehow it still fails exactly the same way on the next test as with NONE flag - by failing to find a device specified in history. May be I am missing something.

Reply:
OK in that regard, I would rather move the cleanup to the start of the function to make this more deterministic.  This is how `fu_history_func` does it as well.

Reply:
I am not sure if that would achieve the necessary effect for the problem comes from the added tests leaving some state which affects running the following one and thus being required to clean after itself, though I am not sure why that is not a problem in some other cases. I guess generically that is a problem of the tests not being hermetic, which is a problem in its own turn, but should be addressed separately.

Reply:
> not being hermetic

Which is why I'm proposing you remove it at the beginning of the test instead.  The other tests that need it will do the same thing.

Reply:
Ah, I see, you mean clean it up in the following test, I misunderstood. Done.